newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js

```diff
@@ -0,0 +1,565 @@
+  resourceExternalEditors: Array<ResourceExternalEditor>,
+|};
+
+type State = {|
+  image: string,
+  imageWidth: number,
+  imageHeight: number,
+|};
+
+export default class Physics2Editor extends React.Component<Props, State> {
+  resourcesLoader: typeof ResourcesLoader;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.resourcesLoader = ResourcesLoader;
+
+    this.state = {
+      image: '',
+      imageWidth: 9,
+      imageHeight: 0,
+    };
+  }
+
+  _setImageSize = (width: number, height: number) => {
+    this.setState({
+      imageWidth: width,
+      imageHeight: height,
+    });
+  };
+
+  _isPolygonConvex(properties: Object) {
+    const vertices = JSON.parse(properties.get('vertices').getValue());
+    // Get edges
+    var edges = [];
+    var v1 = null;
+    var v2 = null;
+    for (var i = 0; i < vertices.length; i++) {
+      v1 = vertices[i];
+      if (i + 1 >= vertices.length) v2 = vertices[0];
+      else v2 = vertices[i + 1];
+      edges.push({ x: v2.x - v1.x, y: v2.y - v1.y });
+    }
+    // Check convexity
+    if (edges.length < 3) return false;
+
+    const zProductIsPositive =
+      edges[0].x * edges[0 + 1].y - edges[0].y * edges[0 + 1].x > 0;
+
+    for (i = 1; i < edges.length - 1; ++i) {
+      var zCrossProduct =
+        edges[i].x * edges[i + 1].y - edges[i].y * edges[i + 1].x;
+      if (zCrossProduct > 0 !== zProductIsPositive) return false;
+    }
+
+    var lastZCrossProduct =
+      edges[edges.length - 1].x * edges[0].y -
+      edges[edges.length - 1].y * edges[0].x;
+    if (lastZCrossProduct > 0 !== zProductIsPositive) return false;
+
+    return true;
+  }
+
+  _isBitEnabled(bitsValue: number, i: number) {
+    return !!(bitsValue & (1 << i));
+  }
+
+  _enableBit(bitsValue: number, i: number, enable: boolean) {
+    if (enable) bitsValue |= 1 << i;
+    else bitsValue &= ~(1 << i);
+    return bitsValue;
+  }
+
+  renderBitProperty(
+    properties: Object,
+    isLayer: boolean,
+    pos: number,
+    spacing: boolean
+  ) {
+    const bitValues = parseInt(
+      properties.get(isLayer ? 'layers' : 'masks').getValue(),
+      10
+    );
+
+    return (
+      <div
+        style={{ width: spacing ? '7.5%' : '5%' }}
+        key={(isLayer ? 'layer' : 'mask') + `${pos}`}
+      >
+        {
+          <Checkbox
+            checked={this._isBitEnabled(bitValues, pos)}
+            onCheck={(e, checked) => {
+              const newValue = this._enableBit(bitValues, pos, checked);
+              this.props.behavior.updateProperty(
+                isLayer ? 'layers' : 'masks',
+                newValue.toString(10),
+                this.props.project
+              );
+              this.forceUpdate();
+            }}
+          />
+        }
+        {spacing && (
+          <div
+            style={{ width: '33%' }}
+            key={(isLayer ? 'layers' : 'masks') + 'space' + pos.toString(10)}
+          />
+        )}
+      </div>
+    );
+  }
+
+  renderNumericProperty(

```

Comment:
Note that instead of doing a renderXXX method, you could extract this function out of the component, call it NumericProperty, and have an object as argument. Let's call this argument... props :)

`function NumericProperty(props: {|
     properties: Object,
     name: string,
     limit: boolean,
     min: number,
     step: number
  |})`

And boom, you have a component :) Well actually, you need to pass project and behavior and a "onUpdate" callback. Then you can do in your render:

```
<NumericProperty 
     properties={properties}
     name={...}
     limit
     min={...}
     step={...}
/>
```

(btw, this is what is **super great** with React. The cost of abstracting something in a component is literally to move a function... this make the barrier to refactoring and abstracting things ultra low).

It's a bit more readable that using `renderXXX` methods. It also allow to move these extracted component in other files if needed (here it's fine to keep them in a single file, colocating components is a good thing as long as they are not too big).